src/sparse_int_set.jl

```diff
@@ -0,0 +1,292 @@
+            append!(b.packed, new_ids)
+            b.reverse[i] = collect(new_ids)
+            b.counters[i] = INT_PER_PAGE
+        else
+            for offset in 1:INT_PER_PAGE
+                if a.reverse[i][offset] == 0
+                    push!(b, INT_PER_PAGE*(i-1) + offset)
+                end
+            end
+        end
+    end
+    return b
+end
+#Can this be optimized?
+complement!(a::SparseIntSet) = copy!(a, complement(a))
+
+<(a::SparseIntSet, b::SparseIntSet) = (a<=b) && !isequal(a,b)
+<=(a::SparseIntSet, b::SparseIntSet) = issubset(a, b)
+
+function findfirst_packed_id(i, s::SparseIntSet)
+    pageid, offset = pageid_offset(s, i)
+    if isassigned(s.reverse, pageid)
+        @inbounds id = s.reverse[pageid][offset]
+        return id
+    end
+    return 0
+end
+
+function cleanup!(s::SparseIntSet)
+    if any(iszero, s.counters)
+        isused = x -> isassigned(s.reverse, x) && s.counters[x] != 0
+        last_page_id = findlast(!iszero, s.counters)
+        if last_page_id === nothing
+            empty!(s.reverse)
+            empty!(s.counters)
+            return s
+        else
+            new_pages    = Vector{Vector{Int}}(undef, last_page_id)
+            new_counters = zeros(Int, last_page_id)
+            for i in eachindex(s.reverse)
+                if isused(i)
+                    new_pages[i] = s.reverse[i]
+                    new_counters[i] = s.counters[i]
+                end
+            end
+            resize!(s.reverse, last_page_id)
+            resize!(s.counters, last_page_id)
+            s.reverse .= new_pages
+            s.counters .= new_counters
+            return s
+        end
+    else
+        return s
+    end
+end
+
+collect(s::SparseIntSet) = s.packed
+
+mutable struct ZippedSparseIntSetIterator{VT,IT}
+    current_id::Int
+    valid_sets::VT
+    shortest_set::SparseIntSet
+    excluded_sets::IT
+    function ZippedSparseIntSetIterator(valid_sets::SparseIntSet...;exclude::NTuple{N, SparseIntSet}=()) where{N}
+        shortest = valid_sets[findmin(map(x->length(x), valid_sets))[2]]
+        new{typeof(valid_sets), typeof(exclude)}(zero(eltype(shortest)), valid_sets, shortest, exclude)
+    end
+end
+
+zip(s::SparseIntSet...;kwargs...) = ZippedSparseIntSetIterator(s...;kwargs...)
+
+@inline length(it::ZippedSparseIntSetIterator) = length(it.shortest_set)
+
+in_excluded(id, it::ZippedSparseIntSetIterator{VT,Tuple{}}) where {VT} = false
+
+function in_excluded(id, it)
+    for e in it.excluded_sets
+        if id in e
+            return true
+        end
+    end
+    return false
+end
+
+@inline function id_tids(it, state)
+    id = it.shortest_set.packed[state]
+    return id, map(x -> findfirst_packed_id(id, x), it.valid_sets)
+end
+
+@propagate_inbounds function iterate(it::ZippedSparseIntSetIterator, state=1)
+    if state > length(it)
+        return nothing
+    end
+    id, tids = id_tids(it, state)
+    while !all(x -> x!=0, tids) || in_excluded(id, it)
```

Comment:
Is this application of De Morgan's laws correct?
```suggestion
    while any(iszero, tids) || in_excluded(id, it)
```